File: bc2025/src/philip_06/RobotPlayer.cpp
#include "RobotPlayer.hpp"
#include "Soldier.hpp"
#include "SoldierState/Build.hpp"
#include "roles/Mopper.hpp"
#include "roles/Splasher.hpp"
#include "roles/Tower.hpp"
#include "utils/Constants.hpp"
#include <iostream>

namespace philip_06 {

Random RobotPlayer::rng;  // Java: = null (java.util.Random -> api.hpp Random)
bool RobotPlayer::rush = false;
bool RobotPlayer::fill = false;
bool RobotPlayer::retreat = false;
bool RobotPlayer::build = false;
bool RobotPlayer::built = false;
int RobotPlayer::aliveTurns = 0;
int RobotPlayer::lastSeenEnemy = 0;
MapLocation RobotPlayer::lastEnemyTower;

void RobotPlayer::run(RobotController rc) {
    if (rc.getType() != UnitType::SOLDIER && rc.getType() != UnitType::MOPPER)
        utils::Constants::setup(rc);
    // Java: if (rng == null) rng = new Random(rc.getID()^6969);  (no null state in C++)
    rng = Random(rc.getID() ^ 6969);
    roles::Mopper::pastLocs = vector<MapLocation>();
    lastSeenEnemy = rc.getRoundNum();
    while (true) {
        if (rc.senseNearbyRobots(-1, opponent(rc.getTeam())).size() > 0)
            lastSeenEnemy = rc.getRoundNum();
        aliveTurns++;
        int valid = 0;
        for (const MapLocation& ruin : rc.senseNearbyRuins(-1)) {
            if (rc.senseRobotAtLocation(ruin).getLocation().isNull()) {
                valid++;
                break;
            } else if (rc.senseRobotAtLocation(ruin).getTeam() == rc.getTeam()) {
                if (rc.canSenseLocation(ruin.add(Direction::all[Direction::SOUTH])) &&
                    rc.canRemoveMark(ruin.add(Direction::all[Direction::SOUTH]))) {
                    rc.removeMark(ruin.add(Direction::all[Direction::SOUTH]));
                }
            }
        }
        if (valid == 0) {
            for (const MapInfo& loc : rc.senseNearbyMapInfos(8))
                if (loc.hasRuin() || loc.isWall()) valid++;
            // Now consider conflict between other SRPs
            for (const MapInfo& loc : rc.senseNearbyMapInfos()) {
                if (isSecondary(loc.getMark())) {
                    if (!rc.canSenseLocation(loc.getMapLocation().add(Direction::all[Direction::NORTH])) ||
                        (rc.canSenseLocation(loc.getMapLocation().add(Direction::all[Direction::NORTH])) &&
                         !rc.senseMapInfo(loc.getMapLocation().add(Direction::all[Direction::NORTH])).hasRuin()))
                        if (loc.getMapLocation().distanceSquaredTo(rc.getLocation()) != 16) valid++;
                }
            }
        }
        // Check that there is actually enough space
        if (valid == 0 && rc.canMark(rc.getLocation())) {
            if (rc.getLocation().x > 1 && rc.getLocation().x < rc.getMapWidth() - 2) {
                if (rc.getLocation().y > 1 && rc.getLocation().y < rc.getMapHeight() - 2) {
                    rc.mark(rc.getLocation(), true);
                }
            }
        }
        for (const MapInfo& loc : rc.senseNearbyMapInfos()) {
            if (loc.hasRuin()) {
                if (rc.canCompleteTowerPattern(UnitType::LEVEL_ONE_PAINT_TOWER, loc.getMapLocation())) {
                    if (rc.getType() == UnitType::SOLDIER && SoldierState::Build::movingTowardsToFinish &&
                        SoldierState::Build::toFinish.back() == loc.getMapLocation()) {
                        SoldierState::Build::toFinish.pop_back();
                        SoldierState::Build::movingTowardsToFinish = false;
                    }
                    rc.completeTowerPattern(UnitType::LEVEL_ONE_PAINT_TOWER, loc.getMapLocation());
                    built = true;
                }
                if (rc.canCompleteTowerPattern(UnitType::LEVEL_ONE_MONEY_TOWER, loc.getMapLocation())) {
                    if (rc.getType() == UnitType::SOLDIER && SoldierState::Build::movingTowardsToFinish &&
                        SoldierState::Build::toFinish.back() == loc.getMapLocation()) {
                        SoldierState::Build::toFinish.pop_back();
                        SoldierState::Build::movingTowardsToFinish = false;
                    }
                    rc.completeTowerPattern(UnitType::LEVEL_ONE_MONEY_TOWER, loc.getMapLocation());
                    built = true;
                }
                if (rc.canCompleteTowerPattern(UnitType::LEVEL_ONE_DEFENSE_TOWER, loc.getMapLocation())) {
                    if (rc.getType() == UnitType::SOLDIER && SoldierState::Build::movingTowardsToFinish &&
                        SoldierState::Build::toFinish.back() == loc.getMapLocation()) {
                        SoldierState::Build::toFinish.pop_back();
                        SoldierState::Build::movingTowardsToFinish = false;
                    }
                    rc.completeTowerPattern(UnitType::LEVEL_ONE_DEFENSE_TOWER, loc.getMapLocation());
                    built = true;
                }
            }
            if (isAlly(loc.getMark()) && isSecondary(loc.getMark())) {
                if (rc.canCompleteResourcePattern(loc.getMapLocation())) {
                    rc.completeResourcePattern(loc.getMapLocation());
                }
            }
        }
        try {
            switch (rc.getType()) {
                case UnitType::SOLDIER: Soldier::run(rc); break;
                case UnitType::MOPPER: roles::Mopper::run(rc); break;
                case UnitType::SPLASHER: roles::Splasher::run(rc); break;
                default: roles::Tower::run(rc); break;
            }
            Clock::yield();
        } catch (const GameActionException& e) {
            std::cout << "GameActionException" << '\n';
            std::cout << e.what() << '\n';
        } catch (const std::exception& e) {
            std::cout << "Exception" << '\n';
            std::cout << e.what() << '\n';
        }
    }
}

}  // namespace philip_06

File: py/bc2025/src/philip_06/RobotPlayer.py
# Port of philip_06/RobotPlayer.java. See py/CONVENTIONS.md.
# Bot-internal imports use the module form (`import X` / `from pkg import X`)
# because this bot's import graph is cyclic - names resolve at call time, like
# the reechee port's `import RobotPlayer` + `RobotPlayer.RobotPlayer.*`.
import traceback

from api import *


class RobotPlayer:
    # @SuppressWarnings("unused")
    rng = None  # Java: java.util.Random rng = null -> api.Random
    rush = False
    fill = False
    retreat = False
    build = False
    built = False
    aliveTurns = 0
    lastSeenEnemy = 0
    lastEnemyTower = None  # Java: declared without initializer -> null

    @staticmethod
    def run(rc):
        if rc.getType() != UnitType.SOLDIER and rc.getType() != UnitType.MOPPER:
            Constants.Constants.setup(rc)
        if RobotPlayer.rng is None:
            RobotPlayer.rng = Random(rc.getID() ^ 6969)
        RobotPlayer.rng = Random(rc.getID() ^ 6969)
        Mopper.Mopper.pastLocs = []  # Java: new ArrayList<>()
        RobotPlayer.lastSeenEnemy = rc.getRoundNum()
        while True:
            # Java: rc.getTeam().opponent() -> api opponent(rc.getTeam())
            if len(rc.senseNearbyRobots(-1, opponent(rc.getTeam()))) > 0:
                RobotPlayer.lastSeenEnemy = rc.getRoundNum()
            RobotPlayer.aliveTurns += 1
            valid = 0
            for ruin in rc.senseNearbyRuins(-1):
                if rc.senseRobotAtLocation(ruin) is None:
                    valid += 1
                    break
                elif rc.senseRobotAtLocation(ruin).getTeam() == rc.getTeam():
                    if rc.canSenseLocation(ruin.add(Direction.SOUTH)) and rc.canRemoveMark(ruin.add(Direction.SOUTH)):
                        rc.removeMark(ruin.add(Direction.SOUTH))
            if valid == 0:
                for loc in rc.senseNearbyMapInfos(8):
                    if loc.hasRuin() or loc.isWall():
                        valid += 1
                # Now consider conflict between other SRPs
                for loc in rc.senseNearbyMapInfos():
                    if loc.getMark().isSecondary():
                        if not rc.canSenseLocation(loc.getMapLocation().add(Direction.NORTH)) or (rc.canSenseLocation(loc.getMapLocation().add(Direction.NORTH)) and not rc.senseMapInfo(loc.getMapLocation().add(Direction.NORTH)).hasRuin()):
                            if loc.getMapLocation().distanceSquaredTo(rc.getLocation()) != 16:
                                valid += 1
            # Check that there is actually enough space
            if valid == 0 and rc.canMark(rc.getLocation()):
                if rc.getLocation().x > 1 and rc.getLocation().x < rc.getMapWidth() - 2:
                    if rc.getLocation().y > 1 and rc.getLocation().y < rc.getMapHeight() - 2:
                        rc.mark(rc.getLocation(), True)
            for loc in rc.senseNearbyMapInfos():
                if loc.hasRuin():
                    if rc.canCompleteTowerPattern(UnitType.LEVEL_ONE_PAINT_TOWER, loc.getMapLocation()):
                        # Java: Build.toFinish.getLast() == loc... is a MapLocation
                        # reference comparison -> `is`
                        if rc.getType() == UnitType.SOLDIER and Build.Build.movingTowardsToFinish and Build.Build.toFinish[-1] is loc.getMapLocation():
                            Build.Build.toFinish.pop()
                            Build.Build.movingTowardsToFinish = False
                        rc.completeTowerPattern(UnitType.LEVEL_ONE_PAINT_TOWER, loc.getMapLocation())
                        RobotPlayer.built = True
                    if rc.canCompleteTowerPattern(UnitType.LEVEL_ONE_MONEY_TOWER, loc.getMapLocation()):
                        if rc.getType() == UnitType.SOLDIER and Build.Build.movingTowardsToFinish and Build.Build.toFinish[-1] is loc.getMapLocation():
                            Build.Build.toFinish.pop()
                            Build.Build.movingTowardsToFinish = False
                        rc.completeTowerPattern(UnitType.LEVEL_ONE_MONEY_TOWER, loc.getMapLocation())
                        RobotPlayer.built = True
                    if rc.canCompleteTowerPattern(UnitType.LEVEL_ONE_DEFENSE_TOWER, loc.getMapLocation()):
                        if rc.getType() == UnitType.SOLDIER and Build.Build.movingTowardsToFinish and Build.Build.toFinish[-1] is loc.getMapLocation():
                            Build.Build.toFinish.pop()
                            Build.Build.movingTowardsToFinish = False
                        rc.completeTowerPattern(UnitType.LEVEL_ONE_DEFENSE_TOWER, loc.getMapLocation())
                        RobotPlayer.built = True
                if loc.getMark().isAlly() and loc.getMark().isSecondary():
                    if rc.canCompleteResourcePattern(loc.getMapLocation()):
                        rc.completeResourcePattern(loc.getMapLocation())
            try:
                t = rc.getType()
                if t == UnitType.SOLDIER:
                    Soldier.Soldier.run(rc)
                elif t == UnitType.MOPPER:
                    Mopper.Mopper.run(rc)
                elif t == UnitType.SPLASHER:
                    Splasher.Splasher.run(rc)
                else:
                    Tower.Tower.run(rc)
                Clock.doYield()
            except GameActionException:
                print("GameActionException")
                traceback.print_exc()

            except Exception:
                print("Exception")
                traceback.print_exc()


# Java's file-level imports are placed below the class: the cyclic bot modules
# (roles/*.py) do `from RobotPlayer import RobotPlayer`, which only resolves
# once the RobotPlayer class object already exists in the partial module.
import Soldier  # Java: same package philip_06, no import statement needed
from SoldierState import Build
from roles import Mopper
from roles import Splasher
from roles import Tower
from utils import Constants
